File: tunnelz/clock.py
import copy
from itertools import izip, tee
from monotonic import monotonic
from .model_interface import ModelInterface, MiProperty, MiModelProperty


class ControllableClock (ModelInterface):
    """A clock with a complete set of controls."""
    # if True, reset the clock's phase to zero on every tap
    retrigger = MiProperty(False, 'set_retrigger')
    one_shot = MiModelProperty('one_shot', 'set_one_shot')

    # time between turning the tick indicator on and then off again, in ms
    min_tick_display_duration = 250

    def __init__(self):
        super(ControllableClock, self).__init__(Clock())
        self.sync = TapSync()
        # keep track of how long it has been since we turned a tick indicator on
        self._tick_age = None

    def initialize(self):
        super(ControllableClock, self).initialize()
        self._set_tick_indicator_off()

    def _set_tick_indicator_off(self):
        self.update_controllers('ticked', False)

    @property
    def curr_angle(self):
        """Proxy the regular clock interface."""
        return self.model.curr_angle

    def tap(self):
        if self.retrigger:
            self.model.reset_on_update = True
        else:
            self.sync.tap()

            # for now, crudely and immediately change the clock rate if we have
            # a new estimate of what it ought to be
            new_rate = self.sync.rate
            if new_rate is not None:
                self.model.rate = new_rate

    def update_state(self, delta_t):
        """Update clock state, and update UI state as well."""
        self.model.update_state(delta_t)

        # if the clock just ticked, reset the tick age counter
        if self.model.ticked:
            self.update_controllers('ticked', True)
            self._tick_age = 0
        # if we're waiting to reset the tick counter
        elif self._tick_age is not None:
            # age the tick time
            self._tick_age += delta_t
            if self._tick_age >= self.min_tick_display_duration:
                self._tick_age = None
                self._set_tick_indicator_off()

    def nudge(self, count):
        """Nudge the phase forward or backward by count/100 of a beat."""
        adjustment = count * (self.model.rate / 100.)
        new_value = self.model.curr_angle + adjustment
        self.model.curr_angle = new_value % 1.0


class TickTimer (object):
    """Determine when to turn on and off a tick indicator.

    Has some time-specific latching behavior to make the ticks longer than a
    single frame.
    """
    # minimum time between tick on and tick off, in ms
    min_duration = 100

    def __init__(self):
        # if None, then we are idling.
        # otherwise, we turned the tick on and we are waiting to turn it off
        # again.
        self.tick_age = None

    def update_state(self, delta_t):
        if self.tick_age is not None:
            self.tick_age += delta_t

    def emit_update(self):
        if self.tick_age is None:
            return None
        elif self.tick_age >= self.min_duration:
            self.tick_age = None
            return False
        else:
            return



class Clock (object):

    def __init__(self):
        self.curr_angle = 0.0
        # in unit angle per second
        self.rate = 0.0

        # did the clock tick on its most recent update?
        self.ticked = True

        # is this clock running in "one-shot" mode?
        # the clock runs for one cycle when triggered then waits for another
        # trigger event
        self.one_shot = False

        # should this clock reset and tick on its next update?
        self.reset_on_update = False

    def update_state(self, delta_t):

        if self.reset_on_update:
            self.ticked = True
            self.curr_angle = 0.0
            self.reset_on_update = False
        else:
            # delta_t has units of ms, need to divide by 1000
            new_angle = self.curr_angle + (self.rate*delta_t/1000.)

            # if we're running in one-shot mode, clamp the angle at 1.0
            if self.one_shot and new_angle >= 1.0:
                self.curr_angle = 1.0
                self.ticked = False
            else:
                # if the phase just escaped our range, we ticked this frame
                self.ticked = new_angle >= 1.0 or new_angle < 0.0
                self.curr_angle = new_angle % 1.0

    def copy(self):
        return copy.copy(self)


class TapSync (object):
    """Handle estimation of phase and rate from a series of taps."""
    # fractional threshold at which we'll discard the current tap buffer and
    # start a new one
    reset_threshold = 0.1

    def __init__(self):
        self._tap_times = []
        self._rate = None
        self._period = None

    @property
    def rate(self):
        return self._rate

    def _reset_buffer(self, tap_time):
        self._tap_times = [tap_time]
        self._rate = None
        self._period = None

    def _add_tap(self, tap_time):
        """Add a tap time to the buffer and update our value estimates if possible."""
        self._tap_times.append(tap_time)
        if len(self._tap_times) > 1:
            # iterate over all pairs of times to get differences
            first_iter, second_iter = tee(self._tap_times)
            next(second_iter, None)
            deltas = list(second - first for first, second in izip(first_iter, second_iter))

            self._period = sum(deltas) / len(deltas)
            self._rate = 1.0 / self._period

    def tap(self):
        """Register a new tap event, and compute a new estimated bpm."""
        now = monotonic()

        # if the tap buffer isn't empty, determine elapsed time from the last
        # tap to this one
        if self._period is not None:
            dt = now - self._tap_times[-1]

            # if this single estimate of tempo is within +-10% of current, use it
            # otherwise, empty the buffer and start over
            fractional_difference = (self._period - dt) / self._period

            if abs(fractional_difference) > 0.1:
                # outlier, empty the buffer
                self._reset_buffer(now)
            else:
                # append to buffer and update
                self._add_tap(now)
        else:
            self._add_tap(now)

File: tunnels/src/clock.rs
use serde::{Deserialize, Serialize};
use std::time::{Duration, Instant};
use tunnels_lib::number::{BipolarFloat, Phase, UnipolarFloat};

use crate::transient_indicator::TransientIndicator;

/// The number of times a clock has ticked.
/// Signed to support negative rates.
pub type Ticks = i64;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Clock {
    /// The current phase of this clock.
    phase: Phase,
    /// The total number of ticks this clock has made.
    ticks: Ticks,
    /// in unit angle per second
    pub rate: f64,
    /// did the clock tick on its most recent update?
    ticked: bool,
    /// is this clock running in "one-shot" mode?
    /// the clock runs for one cycle when triggered then waits for another
    /// trigger event
    one_shot: bool,
    /// should this clock run?
    run: bool,
    /// should this clock reset and tick on the next state update action?
    reset_on_update: bool,
    /// Should this clock scale its rate during update by the audio envelope?
    pub use_audio: bool,
}

impl Default for Clock {
    fn default() -> Self {
        Self::new()
    }
}

impl Clock {
    pub fn new() -> Self {
        Self {
            phase: Phase::ZERO,
            ticks: 0,
            rate: 0.0,
            ticked: true,
            one_shot: false,
            reset_on_update: false,
            run: true,
            use_audio: false,
        }
    }

    pub fn update_state(&mut self, delta_t: Duration, audio_envelope: UnipolarFloat) {
        if self.reset_on_update {
            self.ticked = true;
            self.ticks = 0;
            // Reset phase to zero or one, depending on sign of rate.
            self.phase = if self.rate >= 0.0 {
                Phase::ZERO
            } else {
                Phase::ONE
            };
            self.reset_on_update = false;
            self.run = true;
            return;
        }

        if !self.run {
            return;
        }

        let rate_modulation = if self.use_audio {
            audio_envelope
        } else {
            UnipolarFloat::ONE
        };

        let new_angle =
            self.phase.val() + (self.rate * rate_modulation.val() * delta_t.as_secs_f64());

        // if we're running in one-shot mode, clamp the angle at 1.0
        if self.one_shot && new_angle >= 1.0 {
            self.phase = Phase::ONE;
            self.ticked = false;
            self.run = false;
        } else if self.one_shot && new_angle < 0.0 {
            self.phase = Phase::ZERO;
            self.ticked = false;
            self.run = false;
        } else {
            // if the phase just escaped our range, we ticked this frame
            self.ticked = !(0.0..1.0).contains(&new_angle);
            if self.ticked {
                self.ticks = self.ticks.wrapping_add(new_angle.div_euclid(1.0) as i64);
            }
            self.phase = Phase::new(new_angle);
        }
    }

    fn set_one_shot(&mut self, one_shot: bool) {
        self.one_shot = one_shot;
        if !one_shot {
            self.run = true;
        }
    }

    pub fn phase(&self) -> Phase {
        self.phase
    }

    pub fn ticks(&self) -> Ticks {
        self.ticks
    }
}

/// A static snapshot of externally-visible ControllableClock state.
#[derive(Default, Serialize, Deserialize, Debug, Clone)]
pub struct StaticClock {
    pub phase: Phase,
    pub ticks: Ticks,
    pub submaster_level: UnipolarFloat,
    pub use_audio_size: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
/// A clock with a complete set of controls.
pub struct ControllableClock {
    clock: Clock,
    sync: TapSync,
    tick_indicator: TransientIndicator,
    /// If true, reset the clock's phase to zero on every tap.
    retrigger: bool,
    /// submaster level for this clock
    submaster_level: UnipolarFloat,
    /// If true, modulate the submaster level using audio envelope.
    use_audio_size: bool,
}

impl Default for ControllableClock {
    fn default() -> Self {
        Self::new()
    }
}

impl ControllableClock {
    /// radial units/s, permitting a max internal clock rate of 1.5 Hz
    /// the negative sign is here so that turning the animation speed knob
    /// clockwise makes the animation appear to run around the beam in the same
    /// direction
    pub const RATE_SCALE: f64 = -1.5;

    pub fn new() -> Self {
        Self {
            clock: Clock::new(),
            sync: TapSync::new(),
            tick_indicator: TransientIndicator::new(Duration::from_millis(100)),
            retrigger: false,
            submaster_level: UnipolarFloat::ONE,
            use_audio_size: false,
        }
    }

    /// Return the current phase of this clock.
    pub fn phase(&self) -> Phase {
        self.clock.phase()
    }

    /// Return the number of ticks this clock has ticked.
    pub fn ticks(&self) -> Ticks {
        self.clock.ticks()
    }

    /// Return the current submaster level.
    pub fn submaster_level(&self) -> UnipolarFloat {
        self.submaster_level
    }

    /// Return true if we should use audio envelope to scale submaster level.
    /// This is returned independently, rather than applied to the submaster
    /// level directly, to allow clients of this submaster to avoid double-
    /// modulating with audio envelope.
    pub fn use_audio_size(&self) -> bool {
        self.use_audio_size
    }

    /// Get all clock state bundled into a struct.
    pub fn as_static(&self) -> StaticClock {
        StaticClock {
            phase: self.phase(),
            ticks: self.ticks(),
            submaster_level: self.submaster_level(),
            use_audio_size: self.use_audio_size(),
        }
    }

    /// Update the state of this clock.
    /// The clock may need to emit state update messages.
    pub fn update_state<E: EmitStateChange>(
        &mut self,
        delta_t: Duration,
        audio_envelope: UnipolarFloat,
        emitter: &mut E,
    ) {
        self.clock.update_state(delta_t, audio_envelope);
        if let Some(tick_state) = self.tick_indicator.update_state(delta_t, self.clock.ticked) {
            emitter.emit_clock_state_change(StateChange::Ticked(tick_state));
        }
    }

    /// Emit the current value of all controllable state.
    pub fn emit_state<E: EmitStateChange>(&self, emitter: &mut E) {
        use StateChange::*;
        emitter.emit_clock_state_change(Retrigger(self.retrigger));
        emitter.emit_clock_state_change(OneShot(self.clock.one_shot));
        emitter.emit_clock_state_change(SubmasterLevel(self.submaster_level));
        emitter.emit_clock_state_change(Ticked(self.tick_indicator.state()));
        emitter.emit_clock_state_change(UseAudioSpeed(self.clock.use_audio));
        emitter.emit_clock_state_change(UseAudioSize(self.use_audio_size));
    }

    /// Handle a control event.
    /// Emit any state changes that have happened as a result of handling.
    pub fn control<E: EmitStateChange>(&mut self, msg: ControlMessage, emitter: &mut E) {
        use ControlMessage::*;
        match msg {
            Set(sc) => self.handle_state_change(sc, emitter),
            Tap => {
                if self.retrigger {
                    self.clock.reset_on_update = true;
                } else if let Some(rate) = self.sync.tap() {
                    self.clock.rate = rate;
                    emitter.emit_clock_state_change(StateChange::Rate(BipolarFloat::new(
                        self.clock.rate / ControllableClock::RATE_SCALE,
                    )));
                }
            }
            ToggleOneShot => {
                self.handle_state_change(StateChange::OneShot(!self.clock.one_shot), emitter);
            }
            ToggleRetrigger => {
                self.handle_state_change(StateChange::Retrigger(!self.retrigger), emitter);
            }
            ToggleUseAudioSize => {
                self.handle_state_change(StateChange::UseAudioSize(!self.use_audio_size), emitter);
            }
            ToggleUseAudioSpeed => {
                self.handle_state_change(
                    StateChange::UseAudioSpeed(!self.clock.use_audio),
                    emitter,
                );
            }
        }
    }

    fn handle_state_change<E: EmitStateChange>(&mut self, sc: StateChange, emitter: &mut E) {
        use StateChange::*;
        match sc {
            Rate(v) => self.clock.rate = v.val() * ControllableClock::RATE_SCALE,
            Retrigger(v) => self.retrigger = v,
            OneShot(v) => self.clock.set_one_shot(v),
            SubmasterLevel(v) => self.submaster_level = v,
            UseAudioSpeed(v) => self.clock.use_audio = v,
            UseAudioSize(v) => self.use_audio_size = v,
            Ticked(_) => (),
        };
        emitter.emit_clock_state_change(sc);
    }
}

#[derive(Debug, Clone)]
pub enum StateChange {
    Rate(BipolarFloat),
    Retrigger(bool),
    OneShot(bool),
    SubmasterLevel(UnipolarFloat),
    UseAudioSize(bool),
    UseAudioSpeed(bool),
    /// Outgoing only, no effect as control.
    Ticked(bool),
}

#[derive(Debug, Clone)]
pub enum ControlMessage {
    Set(StateChange),
    Tap,
    ToggleOneShot,
    ToggleRetrigger,
    ToggleUseAudioSize,
    ToggleUseAudioSpeed,
}

pub trait EmitStateChange {
    fn emit_clock_state_change(&mut self, sc: StateChange);
}

#[derive(Debug, Clone, Serialize, Deserialize)]
/// Estimate rate from a series of taps.
struct TapSync {
    #[serde(skip)]
    taps: Vec<Instant>,
    rate: Option<f64>,
    period: Option<Duration>,
}

impl TapSync {
    /// Fractional threshold at which we'll discard the current tap buffer and
    /// start a new one.
    const RESET_THRESHOLD: f64 = 0.1;

    pub fn new() -> Self {
        Self {
            taps: Vec::new(),
            rate: None,
            period: None,
        }
    }

    fn reset_buffer(&mut self, tap: Instant) {
        self.taps.clear();
        self.taps.push(tap);
        self.rate = None;
        self.period = None;
    }

    fn add_tap(&mut self, tap: Instant) {
        self.taps.push(tap);
        if self.taps.len() < 2 {
            return;
        }
        // compute rate if we have at least two taps
        if let (Some(first), Some(last)) = (self.taps.first(), self.taps.last()) {
            let period = (*last - *first) / (self.taps.len() as u32 - 1);
            self.period = Some(period);
            self.rate = Some(1.0 / period.as_secs_f64());
        }
    }

    /// Process a tap event.  Return our new rate estimate if we have one.
    pub fn tap(&mut self) -> Option<f64> {
        let tap = Instant::now();
        // if the tap buffer isn't empty, determine elapsed time from the last
        // tap to this one
        match self.period {
            Some(period) => {
                let dt = tap - *self.taps.last().unwrap();

                // if this single estimate of tempo is within +-10% of current, use it
                // otherwise, empty the buffer and start over
                let abs_difference = if period > dt {
                    period - dt
                } else {
                    dt - period
                };
                let fractional_difference = abs_difference.as_secs_f64() / period.as_secs_f64();

                if fractional_difference > Self::RESET_THRESHOLD {
                    // outlier, empty the buffer
                    self.reset_buffer(tap);
                } else {
                    // append to buffer and update
                    self.add_tap(tap);
                }
            }
            None => self.add_tap(tap),
        }
        self.rate
    }
}
